exec-service-wasmer/src/wasmer_metering.rs

```diff
@@ -0,0 +1,189 @@
+            points_limit_global_index,
+            points_used_global_index,
+        ));
+    }
+}
+
+#[derive(Debug)]
+struct FunctionMetering {
+    accumulated_cost: u64,
+    opcode_cost: Rc<OpcodeCost>,
+    global_indexes: MeteringGlobalIndexes,
+}
+
+impl FunctionMiddleware for FunctionMetering {
+    fn feed<'b>(
+        &mut self,
+        operator: Operator<'b>,
+        state: &mut MiddlewareReaderState<'b>,
+    ) -> Result<(), MiddlewareError> {
+        // Get the cost of the current operator, and add it to the accumulator.
+        // This needs to be done before the metering logic, to prevent operators like `Call` from escaping metering in some
+        // corner cases.
+        self.accumulated_cost += get_opcode_cost(&operator, self.opcode_cost.as_ref()) as u64;
+
+        // Possible sources and targets of a branch. Finalize the cost of the previous basic block and perform necessary checks.
+        match operator {
+            Operator::Loop { .. } // loop headers are branch targets
+            | Operator::End // block ends are branch targets
+            | Operator::Else // "else" is the "end" of an if branch
+            | Operator::Br { .. } // branch source
+            | Operator::BrTable { .. } // branch source
+            | Operator::BrIf { .. } // branch source
+            | Operator::Call { .. } // function call - branch source
+            | Operator::CallIndirect { .. } // function call - branch source
+            | Operator::Return // end of function - branch source
+            => {
+                if self.accumulated_cost > 0 {
+                    state.extend(&[
+                        // Increment the points used counter.
+                        Operator::GlobalGet { global_index: self.global_indexes.points_used().as_u32() },
+                        Operator::I64Const { value: self.accumulated_cost as i64 },
+                        Operator::I64Add,
+                        Operator::GlobalSet { global_index: self.global_indexes.points_used().as_u32()},
+
+                        // Check if out of gas (points_used >= points_limit)
+                        Operator::GlobalGet { global_index: self.global_indexes.points_used().as_u32() },
+                        Operator::GlobalGet { global_index: self.global_indexes.points_limit().as_u32() },
+                        Operator::I64GeU,
+                        Operator::If { ty: WpTypeOrFuncType::Type(WpType::EmptyBlockType) },
+                        // TODO: insert breakpoint out of gas here
+                        Operator::Unreachable,
+                        Operator::End,
+                    ]);
+
+                    self.accumulated_cost = 0;
+                }
+            }
+            _ => {}
+        }
+        state.push_operator(operator);
+
+        Ok(())
+    }
+}
+
+pub fn set_points_limit(instance: &Instance, limit: u64) {
+    instance
+        .exports
+        .get_global(METERING_POINTS_LIMIT)
```

Comment:
This has nothing to do with WASM globals, right?

Reply:
Unfortunately for us, it has. TBD